src/util/alloc/fd_alloc.h

```diff
@@ -0,0 +1,314 @@
+   relatively space and time efficient ... the overhead is ~4 bytes per
+   allocation).  fd_alloc is not particularly optimized when align>~sz
+   and/or large aligments (>~4096B).  While large values for align are
+   supported by fd_alloc_malloc, directly using fd_wksp_alloc is
+   recommended in such cases.
+
+   If an allocation is "large" (align + sz >~ 64KiB for the current
+   implementation), it will be handled by fd_wksp_alloc under the hood.
+   Otherwise, it will be handled by fd_alloc_malloc algorithms (which
+   are ultimately backed by fd_wksp_alloc).  As such, if a small
+   allocation is "new" (e.g. first allocation of a size around sz, an
+   allocation that can't be packed near other existing allocations
+   around that sz, etc), this might also fallback on fd_wksp_alloc.
+   Typically though, after initial allocation and/or program warmup,
+   fd_alloc_malloc calls will be a reasonably fast O(1) lockfree.
+
+   Returns a pointer to the allocation in the local address space on
+   success.  Note that this pointer will a wksp laddr.  As such, it can
+   be converted to a gaddr, passed to other threads in other thread
+   groups, and converted to a wksp laddr in their address spaces, freed
+   via a join to the fd_alloc in that thread group, persisted beyond the
+   lifetime of the calling thread, etc.
+
+   Returns NULL on failure (silent to support HPC usage) or when sz is
+   0.  Reasons for failure include NULL join, invalid align, sz overflow
+   (sz+align>~2^64), no memory available for request (e.g. workspace has
+   insufficient room or is too fragmented). */
+
+void *
+fd_alloc_malloc( fd_alloc_t * join,
+                 ulong        align,
+                 ulong        sz );
+
+/* fd_alloc_free frees the outstanding allocation whose first byte is
+   pointed to by laddr in the caller's local address space.  join is a
+   current local join to the fd_alloc.  The caller promises laddr was
+   allocated by the underlying fd_alloc (but not necessarily on the
+   calling thread or even in this calling process or even by a thread /
+   process that is still running).  Silent for HPC usage (NULL join and
+   NULL laddr are a no-op).
+
+   Like fd_alloc_malloc, if the allocation was large, this will be
+   handled by fd_wksp_free under the hood, which is neither lockfree nor
+   O(1).  If the allocation was small, this will typically be lockfree
+   O(1).  It is possible that, if the amount of outstanding small
+   allocations has reduced signficantly, fd_alloc_free on a small
+   allocation might trigger a fd_wksp_free to free up wksp space for
+   other usage (including uses not through this fd_alloc).
+
+   (It would be possible to implement this less efficiently in space and
+   time such that join didn't need to be passed.  The current design has
+   picked efficiency and consistency with other APIs though.) */
+
+void
+fd_alloc_free( fd_alloc_t * join,
+               void *       laddr );
+
```

Comment:
Also, would love a f_alloc_dump(...) that just prints the allocation tables with size so that I can look for leaks..

Reply:
Eventually we will have leaks and so would want a way to write tests to validate the memory management at a lower level.   This can wait till a future PR

Reply:
With the most recent commit, we are 99% the way there.

The allocator and underlying wksp now have enough mojo to detect any leaks at program termination (e.g. as per the other discussion, a diagnostic can run through the superblocks in the alloc to find stray small allocations as described before and then run through the wksp partition table to look for stray large allocations via the pointer tags), to dump out program memory usage, etc.

As a bonus, since this has no additional costs in time / space / parallel efficiency or what not, the support will just be available all the time.

It can also do party tricks like selective massive free with same algorithmic efficiency single free and classify all wksp allocations efficiently (so can start doing garbage collection likes things too if so inclined ... mixed feelings at best about that).